### PR TITLE
Fix failure(context canceled) in scheduler_perf benchmark

### PR DESCRIPTION
**Pull request by kerthcet**
Scheduler-perf benchmark tests are always failed due to the cancelled context when tearing down

Signed-off-by: Kante Yin <kerthcet@gmail.com>

#### What type of PR is this?
/sig scheduling
/kind bug
/kind failing-test

#### What this PR does / why we need it:

scheduler-perf benchmarks are always failed due to the error:
```
E0105 04:27:16.047290   27011 perf_utils.go:129] Error while deleting Node: client rate limiter Wait returned an error: context canceled
```

see testgrid: https://testgrid.k8s.io/sig-scalability-benchmarks#scheduler-perf

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kerthcet** [comment]
also cc @pohly

File: test/integration/framework/perf_utils.go
Comment (alculquicondor) [inline test/integration/framework/perf_utils.go:128]: Shouldn't callers to `CleanupNodes` pass the appropriate context?
Otherwise maybe it's worth removing `ctx` from the parameters.

Comment (kerthcet) [reply test/integration/framework/perf_utils.go:128]: Didn't find an appropriate context, `b.Cleanup()` will be called when all its subtests complete, I tried to build a context at the top of `BenchmarkPerfScheduling`, but will cause other failures. 
```
F0105 23:56:05.103306   81606 etcd.go:217] unexpected number of goroutines: before: 2 after 6
```
Let's remove the ctx completely to solve the benchmark failure for now.

Comment (alculquicondor) [reply test/integration/framework/perf_utils.go:128]: Yeah, that's better
/lgtm
/approve

**Comment by kerthcet** [comment]
can you approve again? @alculquicondor

**Comment by alculquicondor** [comment]
I don't have ownership for test/integration/framework

/assign @mikedanese

**Comment by kerthcet** [comment]
@aojea can you help to take a look of this? CI benchmark for scheduler_perf is always failed today.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf_test.go:694]: btw: should we mark the test as failed when CleanupNodes returns an error?
If some nodes aren't deleted due to some errors, the next test cases might be affected by those remaining Nodes.

**Comment by sanposhiho** [comment]
The current ctx used to delete nodes is create d here, right?

https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/scheduler_perf_test.go#L644

Is there possibility that b.Cleanup is called **after** runWorkload func finished?: `ctx` is canceled by defer func, then b.Cleanup is called, and that results in context canceled error?

https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/scheduler_perf_test.go#L642

File: test/integration/framework/perf_utils.go
Comment (kerthcet) [reply test/integration/framework/perf_utils.go:128]: Yes, that's what I want at the end, but as I pasted above, we'll have some errors then. So I'd like to merge this first to fix the CI error and then refactor as what we want together with some other works in `mustSetupScheduler` etc..
```
F0105 23:56:05.103306   81606 etcd.go:217] unexpected number of goroutines: before: 2 after 6
```

Make sense?

Comment (sanposhiho) [reply test/integration/framework/perf_utils.go:128]: The scheduler_perf has already had that `unexpected num of goroutines` error. 

```
unexpected number of goroutines: before: 2 after 43
```

https://storage.googleapis.com/kubernetes-jenkins/logs/ci-benchmark-scheduler-perf-master/1610954337838698496/build-log.txt

So, I'm not sure if passing the appropriate `ctx` really caused this error on your laptop.

Comment (pohly) [reply test/integration/framework/perf_utils.go:128]: > So I'd like to merge this first to fix the CI error 

Sure, but why not do it "properly"? Removing the `ctx` parameter from `CleanupNodes` is going in the wrong direction. Just keep that as-is and change the cleanup code:
```
			b.Cleanup(func() {
                                // This cannot use the ctx above because it will be canceled
                                // by the time that this callback is invoked.
				_ = nodePreparer.CleanupNodes(context.Background())
```

Comment (kerthcet) [reply test/integration/framework/perf_utils.go]: @sanposhiho Yes..., it seems like the error is another case. So I'll do more work.

The problem about goroutine leakage is tracked by https://github.com/kubernetes/kubernetes/issues/114890

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:694]: Fixed.

File: test/integration/framework/perf_utils.go
Comment (kerthcet) [reply test/integration/framework/perf_utils.go]: @pohly would you mind to take another look?

File: test/integration/util/util.go
Comment (pohly) [inline test/integration/util/util.go:110]: Both `StartScheduler` and `StartFakePVController` accept a `ctx` now and both set up cancelation. I suspect that this duplication can be avoided by moving the `context.WithCancel(ctx)` into the caller and then removing `WithCancel` and the `ShutdownFunc` return value from these functions.

**Review by pohly — COMMENTED** [review]
Looks good to me, just one suggestion for further cleanup.

File: test/integration/framework/perf_utils.go
Comment (pohly) [reply test/integration/framework/perf_utils.go]: > The scheduler_perf has already had that `unexpected num of goroutines` error.

The test seems to use "start" functions without a corresponding "wait" function, so the caller cannot be certain that all goroutines have really terminated before returning from the test.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (sanposhiho) [inline test/integration/scheduler_perf/scheduler_perf_test.go:651]: Do we need to create a new `ctx` here?  How about creating `ctx` with timeout in L590 and passing it to `runWorkload`?

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:110]: This is for intended, hope to tear down the kube components almost the same time.

Comment (pohly) [reply test/integration/util/util.go:110]: So why not use the same cancellation context for both?

Comment (pohly) [reply test/integration/util/util.go:110]: The function signatures also become simpler.

Comment (kerthcet) [reply test/integration/util/util.go:110]: `Make the function signatures simpler` makes sense to me. Update as suggested.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:651]: Seems like we can but doesn't look better, parentCtx to cancel operations in b.cleanup() and childCtx to cancel normal operations in `runWorkload`. So I'd like to keep them but I swapped their places

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [inline test/integration/scheduler_perf/scheduler_perf_test.go:584]: Move this into the `func(b *testing.B) {` below, then each test has its own output. Right now, multiple different unrelated test cases share the same output.

You may see warnings about "leaked goroutines": that is printed by the test logger when some goroutine logs something after the test has completed. It doesn't panic (like `g.Logf` normally would), so you can ignore it or use it as reminder that waiting for goroutine completion is missing.

Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Updated.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (sanposhiho) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Is there any reason why we cannot pass `parentCtx` to normal operations (instead of childCtx)?

Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: I suppose it's not `cannot` but `better not`, they're for different lifecycle, parentCtx should be cancelled when the test and its subtests completed, but childCtx can be cancelled when we return from `runWorkload`, if we have some goroutines long-running under `runWorkload`, we can stop them ASAP.

Comment (sanposhiho) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: still not sure. `defer cancel()` in L650 to cancel childCtx and `b.Cleanup(cancel)` in L691 to cancel parentCtx is executed almost at the same time, no?

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: I remembered one more aspect of this: ktesting by default uses log level 5 because its main usage is unit testing where detailed logging is more important after a failure than runtime performance. But for performance benchmarks we want a lower log level. Also useful would be command line flags to increase logging and to enable JSON output. This enables testing the runtime overhead of logging.

I did all of that before, but the "contextual logging in scheduler" PR is still pending. The latest incarnation of the scheduler_perf changes are here: https://github.com/kubernetes/kubernetes/pull/111155/files#diff-3ac53aaed5c5c33f714065046108be1c9c1798fafce05276faf5dadbeb4b17f8

@kerthcet: can you perhaps include that change to test/integration/scheduler_perf/main_test.go in your PR? Then the other PR doesn't need to touch test/integration/scheduler_perf anymore at all because you are covering it completely.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: I think no..., we'll tear down the namespace and nodes at first, then we'll close the kube components, api-server and kube-scheduler etc., finally we'll cancel the parentCtx. `b.Cleanup()` is LIFO. In addition, what if `legacyregistry.Reset()` takes a long time? allthough this may not true here...

**Comment by kerthcet** [comment]
kindly ping @pohly

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_perf_test.go:649]: Why do we need a sub context? Can't we just use `parentCtx` in all cases? If so, it should just be renamed to `ctx`.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Agree, result will be more accurate in performance test with lower log level, and I'm not familiar with contextual logging, so I'll follow your advice to include the CL, but considering this is more than fixing the CI error, I'll make it an append commit then we can revert it easily if need. 

Just append the commit for better review now, once LGTM, I'll squash them with separate commits.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:649]: Also discussed with @sanposhiho about this https://github.com/kubernetes/kubernetes/pull/114843#discussion_r1070815242

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Probably log level 2 would be more appropriate, as it's the recommended log level in production.

Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: I agree, log level 2 as default is better.

It may indeed be cleaner to take out all changes related to ktesting and submit those as a separate PR. Just replace `ctx := ktesting.NewTestContext` with `ctx := context.Background()` and the rest remains the same.

Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Addressed all the comments:
1) change the log level to 2
2) squashed with different commits

**Review by pohly — APPROVED** [review]
Looks good to me.

**Review by pohly — APPROVED** [review]
I was waiting whether there were any other opinions about this, but as that doesn't seem to be the case let's merge it.

/lgtm

**Comment by pohly** [comment]
I was starting to base some of my own work (integration tests for dynamic resource allocation) on this branch when I noticed an issue:
```
$ make test-integration WHAT=./test/integration/scheduler_perf ETCD_LOGLEVEL=warn KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
...
I0124 15:25:40.606292  586250 util.go:328] 1000 pods scheduled
E0124 15:25:40.891464  586250 perf_utils.go:129] Error while deleting Node: client rate limiter Wait returned an error: context canceled
E0124 15:25:40.891494  586250 perf_utils.go:129] Error while deleting Node: client rate limiter Wait returned an error: context canceled
...
```

Looks like some cleanup code is still getting the wrong context.

**Comment by pohly** [comment]
Sorry for the confusion - I was on the master branch, not this one. The error above does not occur there anymore.

/lgtm
/hold cancel

**Comment by alculquicondor** [comment]
/hold
I have some questions still

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Do we have goroutines spawning from runWorkload? It's very confusing that within the function we use both parentCtx and ctx.

File: test/integration/scheduler_perf/main_test.go
Comment (pohly) [reply test/integration/scheduler_perf/main_test.go:47]: It's coming into logsapi, but isn't there yet: https://github.com/kubernetes/kubernetes/pull/114731

Comment (alculquicondor) [reply test/integration/scheduler_perf/main_test.go:47]: @kerthcet please add `TODO(#114731)` to make sure to remove this code

Comment (pohly) [reply test/integration/scheduler_perf/main_test.go:47]: Looks like that is getting reviewed sooner than I thought. I still need an API approver, though.

It might make sense to wait for it - up to you. If not, then I'll update it once it both PRs are merged.

Comment (alculquicondor) [reply test/integration/scheduler_perf/main_test.go:47]: Let's wait

Comment (pohly) [reply test/integration/scheduler_perf/main_test.go:47]: @kerthcet: https://github.com/kubernetes/kubernetes/pull/114731 got merged, please replace this section with `logsapi.AddGoFlags(flag.CommandLine)`.

File: test/integration/scheduler_perf/main_test.go
Comment (pohly) [inline test/integration/scheduler_perf/main_test.go:51]: While you are at it, this also would be useful here:
```
	// Contextual logging is needed for per-test output.
	// This can be removed once it's beta and logsapi.ValidateAndApply no
	// longer disables it when no feature gates are provided.
	klog.EnableContextualLogging(true)
```

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Yes, we do, like starting the api-server/kube-scheduler, and api-server should only be shut-down after nodes/pods teared down. So we use parentCtx in all the cleanup.

File: test/integration/scheduler_perf/main_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/main_test.go:47]: When I added the `logsapi.AddGoFlags(flag.CommandLine)`, error happens, see below. Let me cover this commit in another PR and only focus on fixing CI error in this one.
```
/var/folders/c_/7mkl6xtj6qz3dpn1ggg4d2yw0000gn/T/go-build3763591496/b001/scheduler_perf.test flag redefined: log-flush-frequency 
```

Comment (pohly) [reply test/integration/scheduler_perf/main_test.go]: Ah, that's because of the `if flag.CommandLine.Lookup(f.Name) == nil {` check that isn't in `AddGoFlags`. As you said elsewhere, let's deal with this in a separate PR. I can also take over that part if you want.

Comment (kerthcet) [reply test/integration/scheduler_perf/main_test.go]: That would be great Patrick, thanks.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (sanposhiho) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Ok, got what you try to do. 
But, two different `ctx`s in `runWorkload` looks confusing to me as @alculquicondor said. 

So, how about:
- move `mustSetupScheduler` in L660 - L661 out of `runWorkload`. Probably after L593.
- use `defer` instead of `b.Cleanup` in `runWorkload`. They are executed when `runWorkload` is finished, thus before the kube-apiserver/scheduler is shut-downed.
- The child ctx (now created in L647) is no longer needed. Use only passed `parentCtx` in `runWorkload`.

Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Another alternative: do not create a subcontext. Using the parent context should work if there are no go-routines that should end during `runWorkload`

Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Since all of you think this is confusing, I would not derive a context in runWorkload then. Updated.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_perf_test.go:588]: TODO has the connotation that we still have something to fix.

```suggestion
					ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
```

**Review by alculquicondor — COMMENTED** [review]
one nit and one question

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:588]: ... which in this case is true: `context.TODO` should be replaced with a context that has a logger for per-test output. This was done in a separate commit of this PR, but then got taken out because it is not strictly related to it.

I'm fine with keeping the `context.TODO` and will follow up shortly after this PR is merged.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:667]: That is correct.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go:588]: ah, sounds good.

**Pull request merged**
